### ISSUE TITLE
train_ssd7 can not converge with small dataset

### ISSUE THREAD

**Issue by luckyuho** [pierluigiferrari/ssd_keras#30]
I use train_ssd7 to train 67 pictures of udacity dataset provided by you, but it seems can not work like[here](https://github.com/luckyuho/ssd_test/blob/master/train_ssd7_small_dataset.ipynb).

There are something I have changed:
(1) dataset: only 67 pictures as my dataset [here](https://github.com/luckyuho/ssd_test/tree/master/test_data)
(2) learning rate: 0.00001, because 0.01 and 0.0001 keep loss staying large, I reduce learning rate to hope it can learn something at least, but it turns out the same bad result
(3) epoch: 500, because I have little dataset, I think I need to let it train the times of the number of pictures according to the parameter in ssd7
(4) EarlyStopping: I use patience=epochs to turn off it, because I think small dataset can not learn much in each epoch, it needs more epochs to simulate large dataset

I think small dataset would cause over-fitting which can recognize the object provided by my dataset but can not tell objects from other source, however, it seems that it does not learn anything.
Am I wrong?
Thank you!!

**Comment by mikeszabi** [comment]
I am trying to use ssd_7 to detect power pylons  - which have pretty large aspect ratios (height/width). Small datasets (~200 images) would not converge for me too. I wonder it is because of the unusual aspect ratios or because fo the lack of enough training images...
My aspect ratios are: 
aspect_ratios = [2.5, 3.5, 4.5, 5.5]

**Comment by luckyuho** [comment]
hello, pierluigiferrari
I have tried about 2000 images for ssd7, however it still can work but with bad prediction.
I still can not figure why we need so lots data for predicting our training data, so I assume maybe it is caused by the difficult environment and variant objects.
Therefore, I build a environment myself in unity like this.
<img width="1122" alt="unity_scene" src="https://user-images.githubusercontent.com/18595678/33974753-1a0d072e-e0c5-11e7-979d-205319d407a0.png">
I want to set all same types of cars and people, the only different things would be color and scale
What I try to do is recognize road, people and car like below
<img width="1280" alt="labels" src="https://user-images.githubusercontent.com/18595678/33975066-2b8054aa-e0c7-11e7-8028-bfe468c01f74.png">
I want to try an easy dataset first to learn how the network works, but I do not understand the number of data is good to work, all I know is the more the better ==".
so how many data do you think for ssd7 to get work in this environment?
Thank you!!^^

**Comment by pierluigiferrari** [comment]
I've just tried to overfit a small fraction of the Udacity dataset (120 images), and it overfits quite nicely:

![ssd7_120_images_overfitting](https://user-images.githubusercontent.com/17439178/34064544-1353f31c-e1fa-11e7-9244-1ec957a8e942.png)

So I cannot confirm the issue you're having. That being said, I did make a change to the loss function in a recent update just now. Maybe try again with the updated version.

You can also use the new `random_sample` option of `BatchGenerator`'s `parse_csv()`. It allows you to randomly sample a fraction of a dataset. In my overfitting experiment I used 0.005 of the original Udacity dataset.

@mikeszabi the small number of training images are not the cause if it doesn't converge. Fewer images make the weights converge faster (and overfit of course). Be careful with the aspect ratios though: Using large aspect ratios without adjusting the scaling factors at the same time could lead to anchor boxes that are larger than the receptive field of their predictor layer. If you want to do this in a systematic way you would really have to compute what your aspect ratios actually mean in terms of the sizes of the resulting anchor boxes in pixels and compare that to the receptive fields of the respective predictor layers. I know it sounds tedious to do that, but training an SSD is non-trivial in this respect. I recommend reading and understanding the first part of the code of `SSDBoxEncoder`'s `generate_anchor_boxes()` method. The widths and heights of the anchor boxes are computed as follows:

```
w = scale * size * sqrt(aspect_ratio)
h = scale * size / sqrt(aspect_ratio)
```

You might have to decrease the scaling factors in order to get anchor box sizes that make sense for the respective predictor layers. I believe it is likely that decreasing the scaling factors would help.

**Comment by luckyuho** [comment]
Oh, it's my fault!
and random sampling is really important when we only has small dataset!!(my little experiment)
really thanks for your help, pierluigiferrari!!!
